packages/react-async/src/component.tsx

```diff
@@ -1,135 +1,330 @@
-  KeepFreshProps
-> {
-  (props: Props & ConstantProps): React.ReactElement<
-    ComponentProps<typeof Async>
-  >;
-  Preload(props: PreloadProps & ConstantProps): React.ReactElement<{}>;
-  Prefetch(props: PrefetchProps & ConstantProps): React.ReactElement<{}>;
-  KeepFresh(props: KeepFreshProps & ConstantProps): React.ReactElement<{}>;
+  Props extends object,
+  PreloadOptions extends object = {},
+  PrefetchOptions extends object = {},
+  KeepFreshOptions extends object = {}
+> extends ResolverOptions<ComponentType<Props>> {
+  defer?: DeferTiming | ((props: Props) => boolean);
+  deferHydration?: DeferTiming | ((props: Props) => boolean);
+  displayName?: string;
+  renderLoading?(props: Props): ReactNode;
+  renderError?(error: Error): ReactNode;
+  usePreload?(props: PreloadOptions): () => void;
+  usePrefetch?(props: PrefetchOptions): () => void;
+  useKeepFresh?(props: KeepFreshOptions): () => void;
 }
 
 export function createAsyncComponent<
-  Props,
-  PreloadProps = {},
-  PrefetchProps = {},
-  KeepFreshProps = {}
+  Props extends object,
+  PreloadOptions extends object = {},
+  PrefetchOptions extends object = {},
+  KeepFreshOptions extends object = {}
 >({
   id,
   load,
   defer,
-  renderLoading,
-  renderPreload = noopRender,
-  renderPrefetch = noopRender,
-  renderKeepFresh = noopRender,
+  deferHydration,
+  displayName,
+  renderLoading = noopRender,
+  renderError = defaultRenderError,
+  usePreload: useCustomPreload = noopUse,
+  usePrefetch: useCustomPrefetch = noopUse,
+  useKeepFresh: useCustomKeepFresh = noopUse,
 }: Options<
   Props,
-  PreloadProps,
-  PrefetchProps,
-  KeepFreshProps
->): AsyncComponentType<Props, PreloadProps, PrefetchProps, KeepFreshProps> {
-  function AsyncComponent(props: Props & ConstantProps) {
-    // Can't create a spread from union type, so opt for a function that
-    // does the dirty casting for us.
-    const [componentProps, asyncProps] = splitProps(props);
+  PreloadOptions,
+  PrefetchOptions,
+  KeepFreshOptions
+>): AsyncComponentType<
+  ComponentType<Props>,
+  Props,
+  PreloadOptions,
+  PrefetchOptions,
+  KeepFreshOptions
+> {
+  const resolver = createResolver({id, load});
+  const componentName = displayName || displayNameFromId(resolver.id);
+  const deferred = defer != null;
+  const progressivelyHydrated = deferHydration != null;
+  const scriptTiming =
+    deferred || progressivelyHydrated
+      ? AssetTiming.CurrentPage
+      : AssetTiming.Immediate;
+  const stylesTiming = deferred
+    ? AssetTiming.CurrentPage
+    : AssetTiming.Immediate;
+
+  function Async(props: Props) {
+    const {resolved: Component, load, loading, error} = useAsync(resolver, {
+      scripts: scriptTiming,
+      styles: stylesTiming,
+      immediate: deferred == null,
+    });
+
+    const {current: startedHydrated} = useRef(useHydrationManager().hydrated);
+
+    if (error) {
+      return renderError(error);
+    }
+
+    let loadingMarkup: ReactNode | null = null;
```

Comment:
These conditionals are kinda gnarly, I'd really like to see them in their own functions, or at least putting the `let` for `contentMarkup` closer to where it is actually used, as it's hard to tell at a glance whether the first couple branches change it this way.